lib/srv/termmanager.go

```diff
@@ -122,56 +122,36 @@ func (g *TermManager) Write(p []byte) (int, error) {
 }
 
 func (g *TermManager) Read(p []byte) (int, error) {
-	if len(g.remaining) > 0 {
-		n := copy(p, g.remaining)
-		g.remaining = g.remaining[n:]
-		return n, nil
-	}
-
-	q := make(chan struct{})
-	defer close(q)
-	c := make(chan bool)
-	go func() {
-		g.readStateUpdate.L.Lock()
-		defer g.readStateUpdate.L.Unlock()
-
-		for {
-			g.mu.Lock()
-			on := g.on
-			g.mu.Unlock()
-
-			select {
-			case c <- on:
-			case <-q:
-				close(c)
-				return
-			}
-
-			g.readStateUpdate.Wait()
-		}
-	}()
+	g.mu.Lock()
+	on := g.on
+	g.mu.Unlock()
 
-	on := <-c
 	for {
-		if !on {
+		// Loop until data flow is enabled and there is data to return, or the session is terminated.
+		for !on || len(g.remaining) == 0 {
 			select {
 			case <-g.closed:
 				return 0, io.EOF
-			case on = <-c:
-				continue
+			case on = <-g.readStateUpdate:
+			case data := <-g.incoming:
+				g.remaining = append(g.remaining, data...)
 			}
 		}
 
-		select {
-		case <-g.closed:
-			return 0, io.EOF
-		case on = <-c:
+		// ensure that data flow wasn't changed and can't change
+		// until the data has been returned
+		g.mu.Lock()
+		on = g.on
+		if !on {
+			g.mu.Unlock()
 			continue
-		case g.remaining = <-g.incoming:
-			n := copy(p, g.remaining)
-			g.remaining = g.remaining[n:]
-			return n, nil
 		}
+
+		n := copy(p, g.remaining)
+		g.remaining = g.remaining[n:]
+
+		g.mu.Unlock()
```

Comment:
Do you need to do this unlock so late? It serves here only to protect g.remaining, but that is unprotected above and the comment on the field says it is not protected by the mutex. You could just unlock it once before the `if !on` condition.